File: .claude/gen_showcase.py
#!/usr/bin/env python3
"""One-off generator for ui/showcase.slint — the desktop review grid that
tiles every page template at native phone resolution with a pass/fail
verdict toggle. Kept in .claude/ (not shipped); rerun if the page set
changes. Also prints the Rust PAGE_NAMES block for examples/showcase.rs."""
import os, sys

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

# (file-stem, exported-component) in display order. Order MUST match the
# PAGE_NAMES const in examples/showcase.rs.
PAIRS = [
    ("home", "HomePage"), ("settings", "SettingsPage"), ("login", "LoginPage"),
    ("podcast", "PodcastPage"), ("inbox", "InboxPage"), ("profile", "ProfilePage"),
    ("chat", "ChatPage"), ("dashboard", "DashboardPage"), ("music-library", "MusicLibraryPage"),
    ("onboarding", "OnboardingPage"), ("photo-grid", "PhotoGridPage"), ("search-results", "SearchResultsPage"),
    ("checkout", "CheckoutPage"), ("post-detail", "PostDetailPage"), ("map", "MapPage"),
    ("video-player", "VideoPlayerPage"), ("calendar", "CalendarPage"), ("notification-center", "NotificationCenterPage"),
    ("wallet", "WalletPage"), ("weather", "WeatherPage"), ("activity-rings", "ActivityRingsPage"),
    ("order-tracking", "OrderTrackingPage"), ("paywall", "PaywallPage"), ("comments", "CommentsPage"),
    ("task-list", "TaskListPage"), ("news-feed", "NewsArticleFeedPage"), ("boarding-pass", "BoardingPassPage"),
    ("restaurant-menu", "RestaurantMenuPage"), ("leaderboard", "LeaderboardPage"), ("crypto-portfolio", "CryptoPortfolioPage"),
    ("product-detail", "ProductDetailPage"), ("smart-home", "SmartHomePage"), ("timer", "TimerPage"),
    ("email-thread", "EmailThreadPage"), ("ride-share-booking", "RideShareBookingPage"), ("meditation", "MeditationPage"),
    ("form-wizard", "FormWizardPage"), ("account-settings", "AccountSettingsPage"), ("help-center", "HelpCenterPage"),
    ("meal-log", "MealLogPage"), ("video-feed", "VideoFeedPage"), ("job-listing", "JobListingPage"),
    ("trip-itinerary", "TripItineraryPage"), ("currency-converter", "CurrencyConverterPage"), ("habit-tracker", "HabitTrackerPage"),
    ("payment-split", "PaymentSplitPage"), ("playlist-detail", "PlaylistDetailPage"), ("e-reader", "EReaderPage"),
    ("poll-results", "PollResultsPage"), ("group-chat-list", "GroupChatListPage"), ("cart", "CartPage"),
    ("address-book", "AddressBookPage"), ("weekly-meal-plan", "WeeklyMealPlanPage"), ("tv-show-detail", "TVShowDetailPage"),
    ("post-creator", "PostCreatorPage"), ("code-review", "CodeReviewPage"), ("workout-session", "WorkoutSessionPage"),
    ("hotel-booking", "HotelBookingPage"), ("world-clock", "WorldClockPage"), ("medication", "MedicationPage"),
    ("event-detail", "EventDetailPage"), ("write-review", "WriteReviewPage"), ("journal-entry", "JournalEntryPage"),
    ("room-thermostat", "RoomThermostatPage"), ("app-error", "AppErrorPage"), ("app-lock", "AppLockScreen"),
    ("subscriptions", "SubscriptionsPage"), ("expense-report", "ExpenseReportPage"), ("photo-viewer", "PhotoViewerPage"),
    ("game-lobby", "GameLobbyPage"), ("document-scanner", "DocumentScannerPage"), ("transit-departures", "TransitDeparturesPage"),
    ("voice-recorder", "VoiceRecorderPage"), ("payment-methods", "PaymentMethodPage"), ("wifi-settings", "WiFiSettingsPage"),
    ("investment-detail", "InvestmentDetailPage"), ("app-permissions", "AppPermissionsPage"), ("order-history", "OrderHistoryPage"),
    ("reading-list", "ReadingListPage"), ("review-summary", "ReviewSummaryPage"), ("flight-search", "FlightSearchPage"),
    ("gift-card", "GiftCardPage"), ("wordle-puzzle", "WordlePuzzlePage"), ("onboarding-hint", "OnboardingHintPage"),
    ("bug-report", "BugReportPage"), ("calculator", "CalculatorPage"), ("message-composer", "MessageComposerPage"),
    ("tip-jar", "TipJarPage"), ("recipe", "RecipePage"), ("multi-select-list", "MultiSelectListPage"),
    ("signup", "SignupPage"), ("app-store-listing", "AppStoreListingPage"), ("album-detail", "AlbumDetailPage"),
    ("countdown-event", "CountdownEventPage"), ("trending-topics", "TrendingTopicsPage"), ("welcome-splash", "WelcomeSplashPage"),
    ("insurance-claim", "InsuranceClaimPage"), ("country-selector", "CountrySelectorPage"), ("smart-tv-remote", "SmartTVRemotePage"),
    ("pet-adoption", "PetAdoptionPage"), ("carpool-search", "CarpoolSearchPage"), ("media-lockscreen", "MediaPlayerLockscreenPage"),
    ("voting-ballot", "VotingBallotPage"), ("driver-on-the-way", "DriverOnTheWayPage"), ("timezone-converter", "TimezoneConverterPage"),
    ("invoice", "InvoicePage"), ("sleep-tracking", "SleepTrackingPage"), ("grocery-list", "GroceryListPage"),
    ("qr-scanner", "QRScannerPage"), ("live-sports-score", "LiveSportsScorePage"), ("appearance-settings", "AppearanceSettingsPage"),
    ("live-stream", "LiveStreamPage"), ("two-factor-auth", "TwoFactorAuthPage"), ("storage-manager", "StorageManagerPage"),
    ("quiz", "QuizPage"), ("profile-edit", "ProfileEditPage"), ("turn-by-turn-nav", "TurnByTurnNavPage"),
    ("community-forum", "CommunityForumPage"), ("loyalty-card", "LoyaltyCardPage"), ("donation", "DonationPage"),
    ("audiobook-player", "AudiobookPlayerPage"), ("video-call", "VideoCallPage"), ("file-browser", "FileBrowserPage"),
    ("savings-goal", "SavingsGoalPage"), ("doctor-appointment", "DoctorAppointmentPage"), ("stock-watchlist", "StockWatchlistPage"),
    ("referral", "ReferralPage"), ("achievements", "AchievementsPage"), ("parking-session", "ParkingSessionPage"),
    ("budget-overview", "BudgetOverviewPage"), ("voicemail", "VoicemailPage"), ("contact-detail", "ContactDetailPage"),
    ("seat-selection", "SeatSelectionPage"), ("send-money", "SendMoneyPage"), ("net-worth", "NetWorthPage"),
    ("dialer", "DialerPage"), ("lab-results", "LabResultsPage"), ("security-checkup", "SecurityCheckupPage"),
    ("store-locator", "StoreLocatorPage"), ("equalizer", "EqualizerPage"), ("camera-capture", "CameraCapturePage"),
    ("download-manager", "DownloadManagerPage"), ("nutrition-label", "NutritionLabelPage"), ("delivery-driver", "DeliveryDriverPage"),
    ("mortgage-calculator", "MortgageCalculatorPage"),
]

# Sanity checks ----------------------------------------------------------
on_disk = sorted(f[:-6] for f in os.listdir(os.path.join(ROOT, "ui/pages")) if f.endswith(".slint"))
listed = sorted(s for s, _ in PAIRS)
missing = [s for s in on_disk if s not in listed]
extra = [s for s in listed if s not in on_disk]
if missing or extra:
    print(f"MISMATCH  missing-from-script={missing}  not-on-disk={extra}", file=sys.stderr)
    sys.exit(1)
assert len(PAIRS) == len(on_disk), f"{len(PAIRS)} != {len(on_disk)}"

COLS = 3
CELL_W = 412
CELL_H = 944          # 52px header + 892px phone
GAP = 16
PAD = 16
rows = (len(PAIRS) + COLS - 1) // COLS
viewport_w = COLS * CELL_W + (COLS - 1) * GAP + 2 * PAD
viewport_h = rows * CELL_H + (rows - 1) * GAP + 2 * PAD

# Emit ui/showcase.slint -------------------------------------------------
out = []
out.append("// Showcase — desktop review grid.")
out.append("// =====================================================================")
out.append("//")
out.append("// Tiles every page template at native phone resolution (412x892), each")
out.append("// with a pass / fail verdict toggle in a header strip. Run with:")
out.append("//")
out.append("//   cargo run --example showcase --features showcase")
out.append("//")
out.append("// `examples/showcase.rs` loads verdicts from `showcase-verdicts.json`,")
out.append("// and rewrites that file in realtime on every tick / cross. This file")
out.append("// is generated by `.claude/gen_showcase.py` — edit that, not this.")
out.append("")
out.append('import { Theme } from "theme.slint";')
out.append('import { ScrollView } from "std-widgets.slint";')
for stem, comp in PAIRS:
    out.append(f'import {{ {comp} }} from "pages/{stem}.slint";')
out.append("")
out.append("// One review cell: a verdict header strip above the phone screen.")
out.append("// `verdict`: 0 unrated / 1 keep / 2 redo. The phone screen is injected")
out.append("// as `@children` into a fixed 412x892 clipped frame.")
out.append("component ShowcaseCell inherits Rectangle {")
out.append("    in property <string> title;")
out.append("    in property <int> verdict;")
out.append("    callback set-verdict(int);")
out.append("")
out.append(f"    width: {CELL_W}px;")
out.append(f"    height: {CELL_H}px;")
out.append("    background: Theme.surface-1;")
out.append("    border-radius: Theme.radius-md;")
out.append("    border-width: 2px;")
out.append("    border-color: root.verdict == 1 ? Theme.accent-success")
out.append("        : root.verdict == 2 ? Theme.accent-danger")
out.append("        : Theme.surface-variant;")
out.append("    clip: true;")
out.append("")
out.append("    VerticalLayout {")
out.append("        spacing: 0;")
out.append("        // Verdict header.")
out.append("        Rectangle {")
out.append("            height: 52px;")
out.append("            background: Theme.surface-2;")
out.append("            HorizontalLayout {")
out.append("                padding-left: 12px;")
out.append("                padding-right: 8px;")
out.append("                spacing: 6px;")
out.append("                alignment: stretch;")
out.append("                VerticalLayout {")
out.append("                    horizontal-stretch: 1;")
out.append("                    alignment: center;")
out.append("                    Text {")
out.append("                        text: root.title;")
out.append("                        color: Theme.on-surface;")
out.append("                        font-size: 15px;")
out.append("                        font-weight: 800;")
out.append("                        overflow: elide;")
out.append("                    }")
out.append("                }")
out.append("                // Keep (tick).")
out.append("                VerticalLayout {")
out.append("                    alignment: center;")
out.append("                    Rectangle {")
out.append("                        width: 40px;")
out.append("                        height: 40px;")
out.append("                        border-radius: 20px;")
out.append("                        background: root.verdict == 1 ? Theme.accent-success")
out.append("                            : keep-touch.pressed ? Theme.surface-pressed : Theme.surface-1;")
out.append("                        animate background { duration: Theme.motion-fast; }")
out.append("                        Image {")
out.append("                            width: 20px;")
out.append("                            height: 20px;")
out.append("                            x: (parent.width - self.width) / 2;")
out.append("                            y: (parent.height - self.height) / 2;")
out.append('                            source: @image-url("icons/check.svg");')
out.append("                            colorize: root.verdict == 1 ? Theme.on-primary : Theme.muted;")
out.append("                            image-fit: contain;")
out.append("                        }")
out.append("                        keep-touch := TouchArea {")
out.append("                            clicked => { root.set-verdict(root.verdict == 1 ? 0 : 1); }")
out.append("                        }")
out.append("                    }")
out.append("                }")
out.append("                // Redo (cross).")
out.append("                VerticalLayout {")
out.append("                    alignment: center;")
out.append("                    Rectangle {")
out.append("                        width: 40px;")
out.append("                        height: 40px;")
out.append("                        border-radius: 20px;")
out.append("                        background: root.verdict == 2 ? Theme.accent-danger")
out.append("                            : redo-touch.pressed ? Theme.surface-pressed : Theme.surface-1;")
out.append("                        animate background { duration: Theme.motion-fast; }")
out.append("                        Image {")
out.append("                            width: 18px;")
out.append("                            height: 18px;")
out.append("                            x: (parent.width - self.width) / 2;")
out.append("                            y: (parent.height - self.height) / 2;")
out.append('                            source: @image-url("icons/close.svg");')
out.append("                            colorize: root.verdict == 2 ? Theme.on-primary : Theme.muted;")
out.append("                            image-fit: contain;")
out.append("                        }")
out.append("                        redo-touch := TouchArea {")
out.append("                            clicked => { root.set-verdict(root.verdict == 2 ? 0 : 2); }")
out.append("                        }")
out.append("                    }")
out.append("                }")
out.append("            }")
out.append("        }")
out.append("        // Phone screen (412x892), injected by the caller.")
out.append("        Rectangle {")
out.append("            width: 412px;")
out.append("            height: 892px;")
out.append("            clip: true;")
out.append("            @children")
out.append("        }")
out.append("    }")
out.append("}")
out.append("")
out.append("export component Showcase inherits Window {")
out.append('    title: "slint-mobile-components — showcase";')
out.append("    preferred-width: 1340px;")
out.append("    preferred-height: 940px;")
out.append("    background: Theme.background;")
out.append("    // The typeface every tiled page inherits (any Text without its own")
out.append("    // font-family picks this up). Defaults to the library token; "
           "examples/showcase.rs")
out.append("    // overrides it at runtime from the SHOWCASE_FONT env var — no rebuild.")
out.append("    in property <string> app-font: Theme.font-family;")
out.append("    default-font-family: root.app-font;")
out.append("")
out.append("    // Filled in by examples/showcase.rs. `verdicts` and `titles` are")
out.append("    // parallel to the cell order below; `summary` is recomputed on")
out.append("    // every change.")
out.append("    in property <[string]> titles;")
out.append("    in-out property <[int]> verdicts;")
out.append('    in property <string> summary: "Loading…";')
out.append("    // (cell-index, new-verdict) — Rust persists + refreshes summary.")
out.append("    callback verdict-changed(int, int);")
out.append("")
out.append("    VerticalLayout {")
out.append("        spacing: 0;")
out.append("        // Summary bar.")
out.append("        Rectangle {")
out.append("            height: 56px;")
out.append("            background: Theme.surface-1;")
out.append("            drop-shadow-blur: Theme.elevation-1-blur;")
out.append("            drop-shadow-color: Theme.elevation-1-color;")
out.append("            drop-shadow-offset-y: Theme.elevation-1-y;")
out.append("            HorizontalLayout {")
out.append("                padding-left: 20px;")
out.append("                padding-right: 20px;")
out.append("                spacing: 16px;")
out.append("                Text {")
out.append("                    text: root.summary;")
out.append("                    color: Theme.on-surface;")
out.append("                    font-size: 16px;")
out.append("                    font-weight: 800;")
out.append("                    vertical-alignment: center;")
out.append("                    horizontal-stretch: 1;")
out.append("                }")
out.append("                Text {")
out.append('                    text: "tick = keep   ·   cross = redo   ·   saved live to showcase-verdicts.json";')
out.append("                    color: Theme.muted;")
out.append("                    font-size: 13px;")
out.append("                    font-weight: 600;")
out.append("                    vertical-alignment: center;")
out.append("                }")
out.append("            }")
out.append("        }")
out.append("        ScrollView {")
out.append(f"            viewport-width: {viewport_w}px;")
out.append(f"            viewport-height: {viewport_h}px;")
out.append("            VerticalLayout {")
out.append(f"                width: {viewport_w}px;")
out.append(f"                padding: {PAD}px;")
out.append(f"                spacing: {GAP}px;")
out.append("                alignment: start;")

for r in range(rows):
    out.append("                HorizontalLayout {")
    out.append(f"                    spacing: {GAP}px;")
    out.append("                    alignment: start;")
    for c in range(COLS):
        i = r * COLS + c
        if i >= len(PAIRS):
            break
        stem, comp = PAIRS[i]
        out.append("                    ShowcaseCell {")
        out.append(f"                        title: root.titles[{i}];")
        out.append(f"                        verdict: root.verdicts[{i}];")
        out.append(f"                        set-verdict(v) => {{ root.verdict-changed({i}, v); }}")
        out.append(f"                        {comp} {{ width: 412px; height: 892px; }}")
        out.append("                    }")
    out.append("                }")

out.append("            }")
out.append("        }")
out.append("    }")
out.append("}")
out.append("")

with open(os.path.join(ROOT, "ui/showcase.slint"), "w") as f:
    f.write("\n".join(out))

# Emit the Rust PAGE_NAMES block ----------------------------------------
names = ", ".join(f'"{s}"' for s, _ in PAIRS)
rust = "const PAGE_NAMES: &[&str] = &[\n"
line = "    "
for s, _ in PAIRS:
    tok = f'"{s}", '
    if len(line) + len(tok) > 96:
        rust += line.rstrip() + "\n"
        line = "    "
    line += tok
rust += line.rstrip() + "\n];"
with open(os.path.join(ROOT, ".claude/page_names.rs.txt"), "w") as f:
    f.write(rust + "\n")

print(f"wrote ui/showcase.slint  ({len(PAIRS)} cells, {rows} rows, "
      f"viewport {viewport_w}x{viewport_h})")
print("wrote .claude/page_names.rs.txt")

File: xtask/src/gen_showcase.rs
//! Regenerates `ui/showcase.slint` (and the parallel `examples/showcase.rs`
//! PAGE_NAMES hint at `.claude/page_names.rs.txt`) from the
//! [`SHOWCASE_ORDER`] list. The order is significant — `showcase-verdicts.json`
//! is parallel-indexed, so moving a stem in this slice shifts every
//! verdict downstream of it.
//!
//! Page class names (e.g. `MortgageCalculatorPage`, `NewsArticleFeedPage`)
//! are *not* hardcoded; we scan each crate's `.slint` file at runtime for
//! its top-level `export component`. That keeps the order list and the
//! file content the only sources of truth.

use crate::{categories::category_of, workspace_root};
use std::fs;

/// Display order — every cell, top-to-bottom and left-to-right. Adding a
/// new page: append its stem here (and create the .slint file under the
/// matching `crates/pages-<cat>/ui/`). Removing a page: delete it here.
const SHOWCASE_ORDER: &[&str] = &[
    "home", "settings", "login", "podcast", "inbox", "profile", "chat", "dashboard",
    "music-library", "onboarding", "photo-grid", "search-results", "checkout", "post-detail",
    "map", "video-player", "calendar", "notification-center", "wallet", "weather",
    "activity-rings", "order-tracking", "paywall", "comments", "task-list", "news-feed",
    "boarding-pass", "restaurant-menu", "leaderboard", "crypto-portfolio", "product-detail",
    "smart-home", "timer", "email-thread", "ride-share-booking", "meditation", "form-wizard",
    "account-settings", "help-center", "meal-log", "video-feed", "job-listing", "trip-itinerary",
    "currency-converter", "habit-tracker", "payment-split", "playlist-detail", "e-reader",
    "poll-results", "group-chat-list", "cart", "address-book", "weekly-meal-plan",
    "tv-show-detail", "post-creator", "code-review", "workout-session", "hotel-booking",
    "world-clock", "medication", "event-detail", "write-review", "journal-entry",
    "room-thermostat", "app-error", "app-lock", "subscriptions", "expense-report", "photo-viewer",
    "game-lobby", "document-scanner", "transit-departures", "voice-recorder", "payment-methods",
    "wifi-settings", "investment-detail", "app-permissions", "order-history", "reading-list",
    "review-summary", "flight-search", "gift-card", "wordle-puzzle", "onboarding-hint",
    "bug-report", "calculator", "message-composer", "tip-jar", "recipe", "multi-select-list",
    "signup", "app-store-listing", "album-detail", "countdown-event", "trending-topics",
    "welcome-splash", "insurance-claim", "country-selector", "smart-tv-remote", "pet-adoption",
    "carpool-search", "media-lockscreen", "voting-ballot", "driver-on-the-way",
    "timezone-converter", "invoice", "sleep-tracking", "grocery-list", "qr-scanner",
    "live-sports-score", "appearance-settings", "live-stream", "two-factor-auth",
    "storage-manager", "quiz", "profile-edit", "turn-by-turn-nav", "community-forum",
    "loyalty-card", "donation", "audiobook-player", "video-call", "file-browser", "savings-goal",
    "doctor-appointment", "stock-watchlist", "referral", "achievements", "parking-session",
    "budget-overview", "voicemail", "contact-detail", "seat-selection", "send-money",
    "net-worth", "dialer", "lab-results", "security-checkup", "store-locator", "equalizer",
    "camera-capture", "download-manager", "nutrition-label", "delivery-driver",
    "mortgage-calculator",
];

const COLS: usize = 3;
const CELL_W: u32 = 412;
const CELL_H: u32 = 944; // 52 header + 892 phone
const GAP: u32 = 16;
const PAD: u32 = 16;

pub fn run() -> Result<(), Box<dyn std::error::Error>> {
    let root = workspace_root();

    // Scrape (stem -> page class name) from each .slint file. The "page
    // class" is the last `export component XxxPage|XxxScreen inherits`
    // line in the file — by convention, internal helpers come first and
    // the page-level component is the export at the bottom.
    let mut pairs: Vec<(String, String)> = Vec::with_capacity(SHOWCASE_ORDER.len());
    for stem in SHOWCASE_ORDER {
        let cat = category_of(stem);
        let path = root
            .join("crates")
            .join(format!("pages-{cat}"))
            .join("ui")
            .join(format!("{stem}.slint"));
        if !path.exists() {
            return Err(format!(
                "missing page source for stem '{stem}' — expected {}",
                path.display()
            )
            .into());
        }
        let text = fs::read_to_string(&path)?;
        let class = last_page_class(&text).ok_or_else(|| {
            format!(
                "no `export component XxxPage|XxxScreen` found in {}",
                path.display()
            )
        })?;
        pairs.push((stem.to_string(), class));
    }

    let rows = pairs.len().div_ceil(COLS);
    let viewport_w = COLS as u32 * CELL_W + (COLS as u32 - 1) * GAP + 2 * PAD;
    let viewport_h = rows as u32 * CELL_H + (rows as u32 - 1) * GAP + 2 * PAD;

    let showcase_path = root.join("ui").join("showcase.slint");
    fs::write(&showcase_path, emit_showcase(&pairs, rows, viewport_w, viewport_h))?;
    println!(
        "wrote {} ({} cells, {} rows, viewport {}x{})",
        showcase_path.display(),
        pairs.len(),
        rows,
        viewport_w,
        viewport_h
    );

    let names_path = root.join(".claude").join("page_names.rs.txt");
    fs::write(&names_path, emit_page_names_rs(&pairs))?;
    println!("wrote {}", names_path.display());
    Ok(())
}

fn emit_showcase(pairs: &[(String, String)], rows: usize, vw: u32, vh: u32) -> String {
    let mut s = String::with_capacity(64 * 1024);
    s.push_str(HEADER);
    for (stem, comp) in pairs {
        let cat = category_of(stem);
        s.push_str(&format!(
            "import {{ {comp} }} from \"@mobile-pages-{cat}/{stem}.slint\";\n"
        ));
    }
    s.push('\n');
    s.push_str(CELL_COMPONENT);
    s.push_str(SHOWCASE_HEAD_AND_SUMMARY);
    s.push_str(&format!(
        "        ScrollView {{\n            \
             viewport-width: {vw}px;\n            \
             viewport-height: {vh}px;\n            \
             VerticalLayout {{\n                \
                 width: {vw}px;\n                \
                 padding: {PAD}px;\n                \
                 spacing: {GAP}px;\n                \
                 alignment: start;\n",
    ));
    for r in 0..rows {
        s.push_str("                HorizontalLayout {\n");
        s.push_str(&format!("                    spacing: {GAP}px;\n"));
        s.push_str("                    alignment: start;\n");
        for c in 0..COLS {
            let i = r * COLS + c;
            if i >= pairs.len() {
                break;
            }
            let (_, comp) = &pairs[i];
            s.push_str("                    ShowcaseCell {\n");
            s.push_str(&format!("                        title: root.titles[{i}];\n"));
            s.push_str(&format!("                        verdict: root.verdicts[{i}];\n"));
            s.push_str(&format!(
                "                        set-verdict(v) => {{ root.verdict-changed({i}, v); }}\n"
            ));
            s.push_str(&format!(
                "                        {comp} {{ width: 412px; height: 892px; }}\n"
            ));
            s.push_str("                    }\n");
        }
        s.push_str("                }\n");
    }
    s.push_str("            }\n");
    s.push_str("        }\n");
    s.push_str("    }\n");
    s.push_str("}\n");
    s
}

/// Find the *last* `export component XxxPage|XxxScreen inherits …`
/// component in a `.slint` source. Internal helpers come first by
/// convention; the page-level component is the export at the bottom.
fn last_page_class(slint: &str) -> Option<String> {
    let prefix = "export component ";
    let mut hit: Option<String> = None;
    for line in slint.lines() {
        let Some(rest) = line.strip_prefix(prefix) else {
            continue;
        };
        let mut it = rest.split_whitespace();
        let name = match it.next() {
            Some(n) => n,
            None => continue,
        };
        let first = name.chars().next();
        let suffix_ok = name.ends_with("Page") || name.ends_with("Screen");
        let starts_upper = matches!(first, Some(c) if c.is_ascii_uppercase());
        if !suffix_ok || !starts_upper {
            continue;
        }
        if it.next() != Some("inherits") {
            continue;
        }
        hit = Some(name.to_string());
    }
    hit
}

fn emit_page_names_rs(pairs: &[(String, String)]) -> String {
    let mut out = String::from("const PAGE_NAMES: &[&str] = &[\n");
    let mut line = String::from("    ");
    for (stem, _) in pairs {
        let tok = format!("\"{stem}\", ");
        if line.len() + tok.len() > 96 {
            out.push_str(line.trim_end());
            out.push('\n');
            line.clear();
            line.push_str("    ");
        }
        line.push_str(&tok);
    }
    if !line.trim().is_empty() {
        out.push_str(line.trim_end());
        out.push('\n');
    }
    out.push_str("];\n");
    out
}

const HEADER: &str = "// Showcase — desktop review grid.
// =====================================================================
//
// Tiles every page template at native phone resolution (412x892), each
// with a pass / fail verdict toggle in a header strip. Run with:
//
//   cargo run --example showcase --features showcase
//
// `examples/showcase.rs` loads verdicts from `showcase-verdicts.json`,
// and rewrites that file in realtime on every tick / cross. This file
// is generated by `cargo xtask gen-showcase` — edit that, not this.

import { Theme } from \"@mobile-theme/theme.slint\";
import { ScrollView } from \"std-widgets.slint\";
";

const CELL_COMPONENT: &str = "// One review cell: a verdict header strip above the phone screen.
// `verdict`: 0 unrated / 1 keep / 2 redo. The phone screen is injected
// as `@children` into a fixed 412x892 clipped frame.
component ShowcaseCell inherits Rectangle {
    in property <string> title;
    in property <int> verdict;
    callback set-verdict(int);

    width: 412px;
    height: 944px;
    background: Theme.surface-1;
    border-radius: Theme.radius-md;
    border-width: 2px;
    border-color: root.verdict == 1 ? Theme.accent-success
        : root.verdict == 2 ? Theme.accent-danger
        : Theme.surface-variant;
    clip: true;

    VerticalLayout {
        spacing: 0;
        // Verdict header.
        Rectangle {
            height: 52px;
            background: Theme.surface-2;
            HorizontalLayout {
                padding-left: 12px;
                padding-right: 8px;
                spacing: 6px;
                alignment: stretch;
                VerticalLayout {
                    horizontal-stretch: 1;
                    alignment: center;
                    Text {
                        text: root.title;
                        color: Theme.on-surface;
                        font-size: 15px;
                        font-weight: 800;
                        overflow: elide;
                    }
                }
                // Keep (tick).
                VerticalLayout {
                    alignment: center;
                    Rectangle {
                        width: 40px;
                        height: 40px;
                        border-radius: 20px;
                        background: root.verdict == 1 ? Theme.accent-success
                            : keep-touch.pressed ? Theme.surface-pressed : Theme.surface-1;
                        animate background { duration: Theme.motion-fast; }
                        Image {
                            width: 20px;
                            height: 20px;
                            x: (parent.width - self.width) / 2;
                            y: (parent.height - self.height) / 2;
                            source: @image-url(\"@mobile-components/icons/check.svg\");
                            colorize: root.verdict == 1 ? Theme.on-primary : Theme.muted;
                            image-fit: contain;
                        }
                        keep-touch := TouchArea {
                            clicked => { root.set-verdict(root.verdict == 1 ? 0 : 1); }
                        }
                    }
                }
                // Redo (cross).
                VerticalLayout {
                    alignment: center;
                    Rectangle {
                        width: 40px;
                        height: 40px;
                        border-radius: 20px;
                        background: root.verdict == 2 ? Theme.accent-danger
                            : redo-touch.pressed ? Theme.surface-pressed : Theme.surface-1;
                        animate background { duration: Theme.motion-fast; }
                        Image {
                            width: 18px;
                            height: 18px;
                            x: (parent.width - self.width) / 2;
                            y: (parent.height - self.height) / 2;
                            source: @image-url(\"@mobile-components/icons/close.svg\");
                            colorize: root.verdict == 2 ? Theme.on-primary : Theme.muted;
                            image-fit: contain;
                        }
                        redo-touch := TouchArea {
                            clicked => { root.set-verdict(root.verdict == 2 ? 0 : 2); }
                        }
                    }
                }
            }
        }
        // Phone screen (412x892), injected by the caller.
        Rectangle {
            width: 412px;
            height: 892px;
            clip: true;
            @children
        }
    }
}

";

const SHOWCASE_HEAD_AND_SUMMARY: &str = "export component Showcase inherits Window {
    title: \"slint-mobile-components — showcase\";
    preferred-width: 1340px;
    preferred-height: 940px;
    background: Theme.background;
    // The typeface every tiled page inherits (any Text without its own
    // font-family picks this up). Defaults to the library token; examples/showcase.rs
    // overrides it at runtime from the SHOWCASE_FONT env var — no rebuild.
    in property <string> app-font: Theme.font-family;
    default-font-family: root.app-font;

    // Filled in by examples/showcase.rs. `verdicts` and `titles` are
    // parallel to the cell order below; `summary` is recomputed on
    // every change.
    in property <[string]> titles;
    in-out property <[int]> verdicts;
    in property <string> summary: \"Loading…\";
    // (cell-index, new-verdict) — Rust persists + refreshes summary.
    callback verdict-changed(int, int);

    VerticalLayout {
        spacing: 0;
        // Summary bar.
        Rectangle {
            height: 56px;
            background: Theme.surface-1;
            drop-shadow-blur: Theme.elevation-1-blur;
            drop-shadow-color: Theme.elevation-1-color;
            drop-shadow-offset-y: Theme.elevation-1-y;
            HorizontalLayout {
                padding-left: 20px;
                padding-right: 20px;
                spacing: 16px;
                Text {
                    text: root.summary;
                    color: Theme.on-surface;
                    font-size: 16px;
                    font-weight: 800;
                    vertical-alignment: center;
                    horizontal-stretch: 1;
                }
                Text {
                    text: \"tick = keep   ·   cross = redo   ·   saved live to showcase-verdicts.json\";
                    color: Theme.muted;
                    font-size: 13px;
                    font-weight: 600;
                    vertical-alignment: center;
                }
            }
        }
";
